src/java/org/apache/cassandra/index/IndexStatusManager.java

```diff
@@ -183,35 +212,81 @@ public synchronized void propagateLocalIndexStatus(String keyspace, String index
             // Gossiper has not been started. If we attempt to propagate when not started an exception is
             // logged and this causes a number of dtests to fail.
             if (Gossiper.instance.isEnabled())
             {
-                String newStatus = JsonUtils.JSON_OBJECT_MAPPER.writeValueAsString(states);
+                // Versions 5.0.0 through 5.0.2 use a much more bloated format that duplicates keyspace names
+                // and writes full status names instead of their numeric codes. If the minimum cluster version is
+                // unknown or one of those 3 versions, continue to propagate the old format.
+                CassandraVersion minVersion = Gossiper.instance.getMinVersion(1, TimeUnit.SECONDS);
+                String newSerializedStatusMap = shouldWriteLegacyStatusFormat(minVersion) ? JsonUtils.writeAsJsonString(statusMap) 
+                                                                                          : toSerializedFormat(statusMap);
+
                 statusPropagationExecutor.submit(() -> {
                     // schedule gossiper update asynchronously to avoid potential deadlock when another thread is holding
                     // gossiper taskLock.
-                    VersionedValue value = StorageService.instance.valueFactory.indexStatus(newStatus);
+                    VersionedValue value = StorageService.instance.valueFactory.indexStatus(newSerializedStatusMap);
                     Gossiper.instance.addLocalApplicationState(ApplicationState.INDEX_STATUS, value);
                 });
             }
         }
-        catch (Throwable e)
+        catch (Exception e)
         {
             logger.warn("Unable to propagate index status: {}", e.getMessage());
         }
     }
 
+    private static boolean shouldWriteLegacyStatusFormat(CassandraVersion minVersion)
+    {
+        return minVersion == null || (minVersion.major == 5 && minVersion.minor == 0 && minVersion.patch < 3);
+    }
+
+    /**
+     * Serializes as a JSON string the status of the indexes in the provided map.
+     * <p> 
+     * For example, the map...
+     * <pre>
+     * {
+     *     ks1.cf1_idx1=FULL_REBUILD_STARTED,
+     *     ks1.cf1_idx2=FULL_REBUILD_STARTED,
+     *     system.PaxosUncommittedIndex=BUILD_SUCCEEDED
+     * }
+     * </pre>
+     * ...will be converted to the string...
+     * <pre>
+     * {
+     *     "system": {"PaxosUncommittedIndex": 3},
+     *     "ks1": {"cf1_idx1": 1, "cf1_idx2": 1}
+     * }
+     * </pre>
+     */
+    public static String toSerializedFormat(Map<String, Index.Status> indexStatusMap)
+    {
+        Map<String, Map<String, Integer>> serialized = new HashMap<>();
+
+        for (Map.Entry<String, Index.Status> e : indexStatusMap.entrySet())
+        {
+            String keyspace = e.getKey().split("\\.")[0];
+            String index = e.getKey().split("\\.")[1];
```

Comment:
why split the same string twice?  save to a variable?

Reply:
Yeah, this is silly...will change...